presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -73,19 +73,16 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
     {
         boolean decomposable = aggregationNode.isDecomposable(functionRegistry);
 
+        if (!decomposable) {
+            return Result.empty();
+        }
+
         if (aggregationNode.getStep().equals(SINGLE) &&
```

Comment:
that condition seems not to make sense and I'm not sure where it helps actually. It seems redundant to: https://github.com/prestodb/presto/pull/10224/files#diff-b7cdbeaf2c1b99719da7ae1d1bb76ca7R124

The situation is as follows. When SINGLE aggregation with default value is distributed then it needs to be split into FINAL and PARTIAL aggregations which are separated by hash exchange. I thought this if initiates the split, but https://github.com/prestodb/presto/pull/10224/files#diff-b7cdbeaf2c1b99719da7ae1d1bb76ca7R124 does it already.

Simply splitting aggregation into `FINAL` and `PARTIAL` as here doesn't help. Those two aggregations are still within the same stage.

We should add a plan validator that will make sure that FINAL aggregations with default value are separated from PARTIAL aggregation by remote hash exchange

Reply:
Without that, planning fails due to DISTINCT aggs being non-decomposable + SINGLE step by the time it gets here. It may be a real issue with handling of grouping sets with mixed empty and non-empty groupings for this particular combination. I'll analyze this more closely.

Reply:
After taking a closer look, now I remember what this code is trying to do. When there's an aggregation with empty and non-empty grouping sets, if the FINAL runs in a hash-distributed stage, there needs to be a PARTIAL that generates default output so that the empty grouping set can produce its results. Otherwise, if if the final sees no input, it wouldn't know which of the nodes has to produce the output for the empty grouping set.

AddExchanges introduces an Exchange below a SINGLE-STEP, and in the case of empty+non-empty grouping sets, this rule *must* fire so that it splits it into a PARTIAL-FINAL across the exchange. As you described offline, this technically leaves a broken plan between AddExchanges and this rule, which is not ideal. Optimizers should never produce plans that are broken -- they are only allowed to produce unimplementable plans.

Interestingly, for this PR it works out because AddExchanges introduces a GATHER instead of a REPARTITION for the case of multiple DISTINCT. This brings up a few observations:
- This rule should allow a SINGLE agg with empty and non-empty grouping sets as long as it sits on top of a GATHER.
- When the agg has empty grouping set, this PR will produce an undistributed plan, which can hurt parallelism and increase the chances of running out of local memory. This is equivalent to what would happen in classic aggregation query with no group-by clause, and it's how grouping sets used to work in the beginning.
- The mechanism for ensuring default output for hash-distributed computation of grouping sets is complicated. I have some ideas on how to improve this.